earthmover/earthmover.py

```diff
@@ -383,3 +391,155 @@ def test(self, tests_dir: str):
+        # Create packages directory
+        if not os.path.isdir(packages_dir):
+            self.logger.info(
+                f"creating package directory {packages_dir}"
+            )
+            os.makedirs(packages_dir, exist_ok=True)
+
+        # Check for cycles in the package graph
+        if not nx.is_directed_acyclic_graph(self.package_graph):
+            _cycle = nx.find_cycle(self.package_graph)
+            self.error_handler.throw(
+                f"The package graph has a cycle! Installation stopped. Cycle: {_cycle}"
+            )
+            raise
+
+        for package_name in package_subgraph.successors(root_node):
+            package_node = self.package_graph.nodes[package_name]
+            # Install packages if necessary, or retrieve path to package yaml file
+            if install:
+                installed_package_yaml = package_node['package'].install(packages_dir)
+            else:
+                package_node['package'].package_path = os.path.join(packages_dir, package_name)
+                installed_package_yaml = package_node['package'].get_installed_config_file()
+
+            package_configs = JinjaEnvironmentYamlLoader.load_project_configs(installed_package_yaml, params=self.params)
+
+            # Update project parameter defaults from the package, if any
+            for key, val in package_configs.get("parameter_defaults", {}).items():
+                self.params.setdefault(key, val)
+
+            # Prepend package macros to the project macro string. Later macro definitions in the string will overwrite earlier ones 
+            self.macros = package_configs.get("macros", "").strip() + self.macros
+
+            # Load the package yaml and check for additional nested packages
+            package_config = JinjaEnvironmentYamlLoader.load_config_file(installed_package_yaml, params=self.params, macros=self.macros)
+            nested_package_config = self.error_handler.assert_get_key(package_config, 'packages', dtype=dict, required=False, default={})
+
+            # Add nested packages to packages_graph
+            for name, config in nested_package_config.items():
+                nested_package = Package(name, config, earthmover=self)
+                self.package_graph.add_node(name, package=nested_package)
+                self.package_graph.add_edge(package_name, name)
+
+            # Install nested packages by calling this function on a subgraph containing the current package node and its successors
+            if any(True for _ in self.package_graph.successors(package_name)):    
+                nested_package_dir = os.path.join(package_node['package'].package_path, 'packages')
+                nested_package_subgraph = nx.ego_graph(self.package_graph, package_name)
+
+                self.build_package_graph(root_node=package_name, package_subgraph=nested_package_subgraph, packages_dir=nested_package_dir, install=install)
+
+
+    @staticmethod
+    def merge_dicts(dict1, dict2):

```

Comment:
I'd argue this method should be built within YamlMapping, since it will only ever be applied to objects of this class.